// frint-react/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _render = require('./render');

var _render2 = _interopRequireDefault(_render);

var _streamProps = require('./streamProps');

var _streamProps2 = _interopRequireDefault(_streamProps);

var _isObservable = require('./isObservable');

var _isObservable2 = _interopRequireDefault(_isObservable);

var _getMountableComponent = require('./components/getMountableComponent');

var _getMountableComponent2 = _interopRequireDefault(_getMountableComponent);

var _observe = require('./components/observe');

var _observe2 = _interopRequireDefault(_observe);

var _Region = require('./components/Region');

var _Region2 = _interopRequireDefault(_Region);

var _Provider = require('./components/Provider');

var _Provider2 = _interopRequireDefault(_Provider);

var _Region3 = require('./services/Region');

var _Region4 = _interopRequireDefault(_Region3);

var _ReactHandler = require('./handlers/ReactHandler');

var _ReactHandler2 = _interopRequireDefault(_ReactHandler);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  render: _render2.default,
  streamProps: _streamProps2.default,
  isObservable: _isObservable2.default,

  getMountableComponent: _getMountableComponent2.default,
  observe: _observe2.default,
  Region: _Region2.default,
  Provider: _Provider2.default,

  RegionService: _Region4.default,

  ReactHandler: _ReactHandler2.default
};
module.exports = exports['default'];
