// frint-react/lib/handlers/ReactHandler.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getMountableComponent = require('../components/getMountableComponent');

var _getMountableComponent2 = _interopRequireDefault(_getMountableComponent);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

exports.default = {
  setData: function setData(key, value) {
    this.component.setState(_defineProperty({}, key, value));
  },
  setDataWithCallback: function setDataWithCallback(key, value, cb) {
    this.component.setState(_defineProperty({}, key, value), cb);
  },
  getData: function getData(key) {
    return this.component.state[key];
  },
  getProps: function getProps() {
    return this.component.props;
  },
  getProp: function getProp(key) {
    return this.component.props[key];
  },

  getMountableComponent: _getMountableComponent2.default
};
module.exports = exports['default'];
