// frint-component-utils/lib/DefaultHandler.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
/* eslint-disable no-unused-vars */
/* istanbul ignore next */
exports.default = {
  // options
  app: null,
  component: null,

  // lifecycle: creation
  initialize: function initialize() {},
  beforeDestroy: function beforeDestroy() {},


  // data
  getInitialData: function getInitialData() {
    return {};
  },
  setData: function setData(key, value) {},
  setDataWithCallback: function setDataWithCallback(key, value, cb) {
    this.setData(key, value);
    cb();
  },
  getData: function getData(key) {},


  // props
  getProp: function getProp(key) {},
  getProps: function getProps() {},


  // lifecycle: mounting
  beforeMount: function beforeMount() {},
  afterMount: function afterMount() {},


  // lifecycle: re-rendering
  beforeUpdate: function beforeUpdate() {},
  shouldUpdate: function shouldUpdate(nextProps, nextData) {},
  afterUpdate: function afterUpdate() {},


  // other
  getMountableComponent: function getMountableComponent(app) {
    return app.get('component');
  }
};
module.exports = exports['default'];
