// frint-config/lib/externals.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.frint = exports.thirdParties = exports.lodash = exports.rxjs = undefined;

var _webpackRxjsExternals = require('webpack-rxjs-externals');

var _webpackRxjsExternals2 = _interopRequireDefault(_webpackRxjsExternals);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var rxjs = exports.rxjs = [(0, _webpackRxjsExternals2.default)()]; /* eslint-disable func-names */
/**
 * Webpack's `externals` equivalent object,
 * listing dependencies that Frint packages use.
 */
var lodash = exports.lodash = [function (context, request, callback) {
  if (request.startsWith('lodash/')) {
    var subModule = request.split('/')[1];

    return callback(null, {
      root: ['_', subModule],
      commonjs: request,
      commonjs2: request,
      amd: request
    });
  }

  return callback();
}];

// full imports
var thirdParties = exports.thirdParties = [{
  'lodash': {
    root: '_',
    commonjs: 'lodash',
    commonjs2: 'lodash',
    amd: 'lodash'
  }
}, {
  'rxjs': {
    root: 'Rx',
    commonjs: 'rxjs',
    commonjs2: 'rxjs',
    amd: 'rxjs'
  }
}, {
  'react': {
    root: 'React',
    commonjs: 'react',
    commonjs2: 'react',
    amd: 'react'
  }
}, {
  'react-dom': {
    root: 'ReactDOM',
    commonjs: 'react-dom',
    commonjs2: 'react-dom',
    amd: 'react-dom'
  }
}, {
  'prop-types': {
    root: 'PropTypes',
    commonjs: 'prop-types',
    commonjs2: 'prop-types',
    amd: 'prop-types'
  }
}];

var frint = exports.frint = [{
  'frint': {
    root: 'Frint',
    commonjs: 'frint',
    commonjs2: 'frint',
    amd: 'frint'
  }
}, {
  'frint-store': {
    root: 'FrintStore',
    commonjs: 'frint-store',
    commonjs2: 'frint-store',
    amd: 'frint-store'
  }
}, {
  'frint-model': {
    root: 'FrintModel',
    commonjs: 'frint-model',
    commonjs2: 'frint-model',
    amd: 'frint-model'
  }
}, {
  'frint-data': {
    root: 'FrintData',
    commonjs: 'frint-data',
    commonjs2: 'frint-data',
    amd: 'frint-data'
  }
}, {
  'frint-react': {
    root: 'FrintReact',
    commonjs: 'frint-react',
    commonjs2: 'frint-react',
    amd: 'frint-react'
  }
}, {
  'frint-router': {
    root: 'FrintRouter',
    commonjs: 'frint-router',
    commonjs2: 'frint-router',
    amd: 'frint-router'
  }
}, {
  'frint-router-react': {
    root: 'FrintRouterReact',
    commonjs: 'frint-router-react',
    commonjs2: 'frint-router-react',
    amd: 'frint-router-react'
  }
}];

exports.default = rxjs.concat(lodash).concat(thirdParties);
